### PR TITLE
Add --force support for HealthChecks.

### PR DESCRIPTION
**Pull request by G-Harmon**
Without the --force flag, healthchecks will be created, but not modified.
With it, healthchecks will be modified

CC @madhusudancs @csbell @nikhiljindal

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/2)

### CODE REVIEW

File: app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go [hunk 1]
```diff
@@ -37,18 +39,69 @@ func TestEnsureHealthCheck(t *testing.T) {
...
 	// Verify that GET does not return NotFound.
 	if _, err := hcp.GetHealthCheck(hcName); err != nil {
 		t.Fatalf("expected nil error, actual: %v", err)
 	}
+
+	force = false
+	err = hcs.EnsureHealthCheck(lbName, []sp.ServicePort{
+		{
+			Port:     port,
+			Protocol: "HTTPS", /* a different protocol */
```
Comment (csbell) [inline app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go]: what does 'a different protocol mean'? only that it's not http?

File: app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go [hunk 2]
```diff
@@ -37,18 +39,69 @@ func TestEnsureHealthCheck(t *testing.T) {
...
+	}, force)
 	if err != nil {
-		t.Fatalf("expected no error in ensuring health check, actual: %v", err)
+		t.Fatalf("expected no error in ensuring health check (force=true), actual: %v", err)
 	}
 	// Verify that GET does not return NotFound.
 	if _, err := hcp.GetHealthCheck(hcName); err != nil {
 		t.Fatalf("expected nil error, actual: %v", err)
 	}
+
```
Comment (csbell) [inline app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go]: Lines 56-84 I would do a range over:
var healthChecks = [] struct {
  force bool
  protocol string
}{
   { false, "HTTPS" },
   { true, "HTTP" },
}

File: app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go [hunk 3]
```diff
@@ -37,18 +39,69 @@ func TestEnsureHealthCheck(t *testing.T) {
...
+	check2 := check
+	check2.Description = "foo"
+	if healthCheckMatches(&check, &check2) {
+		t.Errorf("Want healthCheckMatches(c, c2) = false, c.description differs. got true.")
+	}
+	check.Description = "foo"
+	if !healthCheckMatches(&check, &check2) {
+		t.Errorf("fail")
+	}
+	check2.CreationTimestamp = "1234"
```
Comment (csbell) [inline app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go]: Maybe add comment /* Test for members that should be ignored */

File: app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go (hunk 1 above)
Comment (G-Harmon) [reply app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go]: yes...

**Comment by G-Harmon** [comment]
Review status: 0 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go, line 56 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QFw4Xo0rASQYe--r1-56:-KwxKGoY18L7MQ2BJMCU:bncom1) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/6cdd8dec20007fde9409bcdd4b28786f12e16b8c/app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go#L56)):*
<details><summary><i>Previously, csbell (Christian Bell) wrote…</i></summary><blockquote>

Lines 56-84 I would do a range over:
var healthChecks = [] struct {
  force bool
  protocol string
}{
   { false, "HTTPS" },
   { true, "HTTP" },
}
</blockquote></details>

It looks much better with a table. And the table includes a description field for the error message.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go, line 61 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QFw4Xo0rASQYe--r1-61:-KwxOLcQnqCcdj_hOPTr:b-8wnbg6) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/6cdd8dec20007fde9409bcdd4b28786f12e16b8c/app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go#L61)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

yes...
</blockquote></details>

Done. Improved comment.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go, line 103 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QFw4Xo0rASQYe--r1-103:-KwxK3fsbDKSqJ3xiKLY:b-gjn8q1) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/6cdd8dec20007fde9409bcdd4b28786f12e16b8c/app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go#L103)):*
<details><summary><i>Previously, csbell (Christian Bell) wrote…</i></summary><blockquote>

Maybe add comment /* Test for members that should be ignored */
</blockquote></details>

I added a comment and improved the t.Error message.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2)*

File: app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go [hunk 4]
```diff
@@ -80,7 +80,8 @@ func TestCreateLoadBalancer(t *testing.T) {
 		}
 		return true, ret, nil
 	})
-	if err := lbc.CreateLoadBalancer(&ing); err != nil {
+	force := true
```
Comment (nikhiljindal) [inline app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go]: We should test with force=false by default.
force=true can mask errors.

Also we dont really need this extra var here
If its for documentation, you can write it as `CreateLoadBalancer(&ing, false /* force */)`

**Comment by G-Harmon** [comment]
Reviewed 6 of 7 files at r1, 1 of 1 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[app/mci/cmd/create.go, line 80 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QEkJa8gxVJ6dYQ-r2-80:-Kx9E9fSvDzaQrBRugp-:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/cmd/create.go#L80)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

How about calling this ForceUpdate?
</blockquote></details>

Done.

---

*[app/mci/cmd/create.go, line 110 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QEkJa8gxVJ6dYQ-r2-110:-Kx9DY3f5enVZC8NqGxR:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/cmd/create.go#L110)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Second argument should be force (instead of Force). flag names are all lowercase.
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer.go, line 97 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QEkJa8gxVJ6dYS-r2-97:-Kx9FBV16kxrYggS5z5F:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/healthcheck/healthchecksyncer.go#L97)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

We can remove this TODO now. --force is explicit permission
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer.go, line 101 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QEkJa8gxVJ6dYS-r2-101:-Kx9ZDvOVrayI4n54B1J:b-xwndea) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/healthcheck/healthchecksyncer.go#L101)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Am wondering if we should print the diff and ask the user yes/no instead of throwing an error here.

So if --force is true, then we overwrite all resources without requiring additional user input.
If --force is false or not set, then user needs to explicitly say yes for each resource to be overwritten.

What do you think?

cc @csbell 
</blockquote></details>

I think we should support mci in a scripted environment, so if we want to ask for user input, we should probably have a flag that turns off the question and assumes "no".
Either way, I think we can put it in a separate PR.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer.go, line 120 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QEkJa8gxVJ6dYS-r2-120:-Kx9FXvn9zV50IhPW5ai:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/healthcheck/healthchecksyncer.go#L120)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Yes
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer.go, line 123 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QEkJa8gxVJ6dYS-r2-123:-Kx9FhUuyLs5kYEkDfcd:bkdv0p) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/healthcheck/healthchecksyncer.go#L123)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

selfLink is what we refer to from Backend service, so lets not ignore it :)
</blockquote></details>

Actually, as we discussed in person, we can't check for this because it's an output (and not set in desiredHC).

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer.go, line 130 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QEkJa8gxVJ6dYS-r2-130:-Kx9GB3RLhHWvndzrI_o:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/healthcheck/healthchecksyncer.go#L130)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Should these be glog.V(2)'s?
The parent method already hasPrintf's for is it attempting Update or not.
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go, line 55 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QFw4Xo0rASQYe--r2-55:-Kx9HCkwKePdGd9Qf51G:bgvyzlp) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go#L55)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

How about same health check with force=false?
</blockquote></details>

Done. I hadn't written that one before because it's not exercising any different branches.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go, line 69 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QFw4Xo0rASQYe--r2-69:-Kx9GYUnZbcsoVtuTuUB:b-tjhl0j) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go#L69)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

I assume you wanted to set force=true?
</blockquote></details>

Done. Oops, I was testing the failure message formatting.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go, line 97 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QFw4Xo0rASQYe--r2-97:-Kx9I00zJ5k220exk25D:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/healthcheck/healthchecksyncer_test.go#L97)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

Can remove this TODO now
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 83 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QFw4Xo0rASQYe2-r2-83:-Kx9YS9rEg5QkBLf26rT:b-2uhdnr) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L83)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

We should test with force=false by default.
force=true can mask errors.

Also we dont really need this extra var here
If its for documentation, you can write it as `CreateLoadBalancer(&ing, false /* force */)`
</blockquote></details>

Removed var. 
For documentation, I hoped we could use the style that C++'s clang validates:
https://clang.llvm.org/extra/clang-tidy/checks/misc-argument-comment.html
a(/*foo=*/ true)
But it looks like the go formatter doesn't like that. oh well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2)*

**Comment by nikhiljindal** [comment]
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[app/mci/pkg/gcp/healthcheck/healthchecksyncer.go, line 101 at r2](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2#-Kwvb7QEkJa8gxVJ6dYS-r2-101:-Kx9g77xzcVnRzWMDX56:b-elbg5r) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/255d9b7e5ec2836da94854264cf49c71c6c06f17/app/mci/pkg/gcp/healthcheck/healthchecksyncer.go#L101)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

I think we should support mci in a scripted environment, so if we want to ask for user input, we should probably have a flag that turns off the question and assumes "no".
Either way, I think we can put it in a separate PR.
</blockquote></details>

The scenario I have in mind is that when user wants to fix just one resource, lets say backend service (the existing backend service does not match the desired one).
By default, running create will throw an error.
But running it with --force=true has the potential to overwrite other resources as well.
User wants to overwrite backend service only and we do not have a way to do that.
Asking explicit user consent for each resource that needs update is one way to solve this.

Whatever way we choose, fixing it in a separate PR sounds good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/2)*

**Comment by nikhiljindal** [comment]
/lgtm.
Merging

**Pull request merged**